src/Castle.Core/DynamicProxy/CustomAttributeInfo.cs

```diff
@@ -0,0 +1,301 @@
+// Copyright 2004-2016 Castle Project - http://www.castleproject.org/
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.DynamicProxy
+{
+	using System;
+	using System.Collections.Generic;
+	using System.Linq;
+	using System.Linq.Expressions;
+	using System.Reflection;
+	using System.Reflection.Emit;
+
+	public class CustomAttributeInfo : IEquatable<CustomAttributeInfo>
+	{
+		// Cached empty arrays to avoid unnecessary allocations
+		private static readonly PropertyInfo[] EmptyProperties = new PropertyInfo[0];
+		private static readonly FieldInfo[] EmptyFields = new FieldInfo[0];
+		private static readonly object[] EmptyValues = new object[0];
+
+		private static readonly IEqualityComparer<object> ValueComparer = new AttributeArgumentValueEqualityComparer();
+
+		private readonly CustomAttributeBuilder builder;
+		private readonly ConstructorInfo constructor;
+		private readonly object[] constructorArgs;
+		private readonly IDictionary<string, object> properties;
+		private readonly IDictionary<string, object> fields;
+
+		public CustomAttributeInfo(
+			ConstructorInfo constructor,
+			object[] constructorArgs,
+			PropertyInfo[] namedProperties,
+			object[] propertyValues,
+			FieldInfo[] namedFields,
+			object[] fieldValues)
+		{
+			// Will take care of validating the arguments
+			this.builder = new CustomAttributeBuilder(constructor, constructorArgs, namedProperties, propertyValues, namedFields, fieldValues);
+
+			this.constructor = constructor;
+			this.constructorArgs = constructorArgs;
```

Comment:
That was my thought at first as well. Then I had a second thought. It'll probably just reveal my inadequate grasp of what's going on here, but may not the other parameters (or even individual args of constructorArgs) be single-dimensional arrays, which could be mutated after the fact?